### PR TITLE
Add option to detect malformed byte sequences

### PR DESCRIPTION
**Pull request by 06romix**
This PR adds the option for detecting malformed byte sequences according to #24

### CODE REVIEW

File: src/Dom/Document.php [hunk 1]
```diff
@@ -2145,4 +2147,22 @@ private function addHttpEquivCharset($html)
...
+     *
+     * @param string $source The HTML fragment string.
+     * @throws \InvalidArgumentException if source contains some malformed byte sequences
+     */
+    private function assertNoneMalformedByteSequences($source)
+    {
+        if ($this->options[Option::ADDITIONAL_VALIDATION] && ! mb_check_encoding($source)) {
+            throw new \InvalidArgumentException(
+                'Passed html contains some malformed byte sequences. ' .
+                'Perhaps you use function "substr" instead of "mb_substr".'
```
Comment (schlessera) [inline src/Dom/Document.php:2164]:
```suggestion
                'Provided HTML contains invalid byte sequences. ' .
                'This is usually fixed by replacing string manipulation functions with their `mb_*` multibyte counterparts.'
```

File: src/Dom/Document.php [hunk 2]
```diff
@@ -2145,4 +2147,22 @@ private function addHttpEquivCharset($html)
...
+     *
+     * If passed bad byte sequences DOMDocument would fail in a random way (silently) and produces garbage.
+     *
+     * @param string $source The HTML fragment string.
+     * @throws \InvalidArgumentException if source contains some malformed byte sequences
+     */
+    private function assertNoneMalformedByteSequences($source)
+    {
+        if ($this->options[Option::ADDITIONAL_VALIDATION] && ! mb_check_encoding($source)) {
+            throw new \InvalidArgumentException(
```
Comment (schlessera) [inline src/Dom/Document.php:2162]: This should use a custom exception `InvalidByteSequence` with a named constructor (see `src/Exception`).

**Review by schlessera — CHANGES_REQUESTED** [review]
I think the `mb_check_encoding()` will detect any invalid byte sequence, so the language used should move from "malformed" (which is a specific issue) to "invalid" (which covers all detected issues).

I've provided change suggestions for the code, but the tests will also needed to be adapted to match the adapted code.

File: src/Dom/Document.php (hunk 2 above)
Comment (06romix) [reply src/Dom/Document.php]: Should the exception message contain an HTML source?
If not, is there necessary to write a test for InvalidByteSequence that contains only a message?

**Comment by 06romix** [comment]
I'm with you on that, "invalid" is a good choice for `mb_check_encoding()`.

Thank you for your detailed review. I appreciate it.

File: src/Dom/Document.php (hunk 2 above)
Comment (schlessera) [reply src/Dom/Document.php]: Normally yes, but for string-based HTML, it is actually hard to produce something useful for an exception message that doesn't flood the screen/logs... Not sure how we would go about that here.

Comment (06romix) [reply src/Dom/Document.php]: The third way is to add a second argument to the named constructor and then provide a method like `$invalidByteSequence->getHtmlSource()` to get the html contained invalid byte sequence.

File: tests/Dom/DocumentTest.php [hunk 3]
```diff
@@ -1280,4 +1281,34 @@ public function testCreateElementWithAttributes()
             $document->saveHTMLFragment($element)
         );
     }
+
+    /**
+     * Test asserting none malformed byte sequences
+     *
+     * @covers \AmpProject\Dom\Document::detectInvalidByteSequences()
```
Comment (schlessera) [inline tests/Dom/DocumentTest.php:1288]:
```suggestion
     * @covers \AmpProject\Dom\Document::detectInvalidByteSequences()
     * @covers \AmpProject\Exception\InvalidByteSequence::forHtml()
```

File: tests/Dom/DocumentTest.php [hunk 4]
```diff
@@ -1280,4 +1281,34 @@ public function testCreateElementWithAttributes()
...
+        $document::fromHtmlFragment(
+            "<div>Some text \x00\x81<div>",
+            [Option::CHECK_ENCODING => true]
+        );
+    }
+
+    /**
+     * Test asserting none bad byte sequences
+     *
+     * @covers \AmpProject\Dom\Document::detectInvalidByteSequences()
```
Comment (schlessera) [inline tests/Dom/DocumentTest.php:1303]:
```suggestion
     * @covers \AmpProject\Dom\Document::detectInvalidByteSequences()
     * @covers \AmpProject\Exception\InvalidByteSequence::forHtml()
```

**Review by schlessera — CHANGES_REQUESTED** [review]

**Comment by schlessera** [comment]
Thanks for the PR, @06romix !

**Pull request merged**